hi-ml-histopathology/src/histopathology/utils/report_utils.py

```diff
@@ -0,0 +1,339 @@
+    ax.set_ylim(-.05, 1.05)
+    ax.grid(color='0.9')
+
+
+def _plot_crossval_roc_and_pr_curves(crossval_dfs: Dict[int, pd.DataFrame],
+                                     roc_ax: Axes, pr_ax: Axes) -> None:
+    for k, tiles_df in crossval_dfs.items():
+        slides_groupby = tiles_df.groupby(ResultsKey.SLIDE_ID)
+        labels = slides_groupby[ResultsKey.TRUE_LABEL].agg(pd.Series.mode)
+        scores = slides_groupby[ResultsKey.PROB].agg(pd.Series.mode)
+        plot_roc_curve(labels, scores, label=f"Fold {k}", ax=roc_ax)
+        plot_pr_curve(labels, scores, label=f"Fold {k}", ax=pr_ax)
+    legend_kwargs = dict(edgecolor='none', fontsize='small')
+    roc_ax.legend(**legend_kwargs)
+    pr_ax.legend(**legend_kwargs)
+    format_pr_or_roc_axes('roc', roc_ax)
+    format_pr_or_roc_axes('pr', pr_ax)
+
+
+def plot_crossval_roc_and_pr_curves(crossval_dfs: Dict[int, pd.DataFrame]) -> Figure:
+    """Plot ROC and precision-recall curves for multiple cross-validation runs.
+
+    This will create a new figure with two subplots (left: ROC, right: PR).
+
+    :param crossval_dfs: Dictionary of dataframes with cross-validation indices as keys,
+        as returned by :py:func:`collect_crossval_outputs()`.
+    :return: The created `Figure` object.
+    """
+    fig, axs = plt.subplots(1, 2, figsize=(8, 4))
+    _plot_crossval_roc_and_pr_curves(crossval_dfs, roc_ax=axs[0], pr_ax=axs[1])
+    return fig
+
+
+def get_crossval_metrics_table(metrics_df: pd.DataFrame,
+                               metrics_list: Sequence[str]) -> pd.DataFrame:
+    """Format raw cross-validation metrics into a table with a summary "Mean ± Std" column.
+
+    Note that this function only supports scalar metrics. To format metrics that are logged
+    hroughout training, you should call :py:func:`get_best_epoch_metrics()` first.
+
+    :param metrics_df: Metrics dataframe, as returned by :py:func:`collect_crossval_metrics()` and
+        :py:func:`~health_azure.aggregate_hyperdrive_metrics()`.
+    :param metrics_list: The list of metrics to include in the table.
+    :return: A dataframe with the values of the selected metrics formatted as strings, including a
+        header and a summary column.
+    """
+    header = ["Metric"] + [f"Split {k}" for k in metrics_df.columns] + ["Mean ± Std"]
+    metrics_rows = []
+    for metric in metrics_list:
+        values: pd.Series = metrics_df.loc[metric]
+        mean = values.mean()
+        std = values.std()
+        row = [metric] + [f"{v:.3f}" for v in values] + [f"{mean:.3f} ± {std:.3f}"]
+        metrics_rows.append(row)
+    table = pd.DataFrame(metrics_rows, columns=header).set_index(header[0])
+    return table
+
+
+def get_best_epochs(metrics_df: pd.DataFrame, primary_metric: str, maximise: bool = True) -> Dict[int, int]:
+    """Determine the best epoch for each cross-validation run based on a given metric.
+
+    The returned epoch indices are relative to the logging frequency of the chosen metric, i.e.
```

Comment:
This looks like it could be important to log as a warning

Reply:
Although, this function has no concept of training or validation, so maybe this warning should be moved up to a level where these concepts are relevant?